src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf8StringMarshaller.cs

```diff
@@ -37,32 +37,33 @@ public Utf8StringMarshaller(string? str)
         /// </remarks>
         public Utf8StringMarshaller(string? str, Span<byte> buffer)
         {
-            _allocated = null;
+            _allocated = false;
+
             if (str is null)
             {
-                _span = default;
+                _nativeValue = null;
                 return;
             }
 
+            const int MaxUtf8BytesPerChar = 3;
+
             // + 1 for null terminator
-            int maxByteCount =  Encoding.UTF8.GetMaxByteCount(str.Length) + 1;
-            if (buffer.Length >= maxByteCount)
-            {
-                int byteCount = Encoding.UTF8.GetBytes(str, buffer);
-                buffer[byteCount] = 0; // null-terminate
-                _span = buffer;
-            }
-            else
+            int maxByteCount = checked(MaxUtf8BytesPerChar * str.Length + 1);
+            if (maxByteCount > buffer.Length)
             {
-                _allocated = (byte*)Marshal.AllocCoTaskMem(maxByteCount);
-                int byteCount;
-                fixed (char* ptr = str)
+                // Calculate accurate byte count when the provided stack-allocated buffer is not sufficient
```

Comment:
The built-in marshallers use exact-sized buffers for some cases and conservatively-sized buffer for other cases. For example, check https://github.com/dotnet/runtime/blob/041933fd086f69f839cf38733dd56a1ea9f09866/src/coreclr/System.Private.CoreLib/src/System/StubHelpers.cs#L94-L96 . 

I think we should just do the exact-sized buffers everywhere. The long strings are rare. When the long strings do show up, the 3x over-allocation for them feels like a lot.

Reply:
I am wondering whether it would make sense to fix the `Marshal` methods do the exact-sized allocations as well.